### ISSUE TITLE
cannot uninstall broken formulae because brew validates the formula file on uninstall

### ISSUE THREAD

**Issue by josharian** [Homebrew/brew#7110]
Please see https://github.com/Homebrew/homebrew-core/issues/51073; it contains all of the information requested in the issue template and relevant history.

@Bo98 [suggested I raise the issue encountered there in this repository](https://github.com/Homebrew/homebrew-core/issues/51073#issuecomment-594192303).

## What you were trying to do (and why)

Uninstall a broken formula. `$ brew uninstall riscv-pk`.

## What happened (include command output)

<details>
  <summary>Command output</summary>
  <pre>

$ brew uninstall riscv-pk 
Error: invalid attribute for formula 'riscv-pk': version (nil)
Please report this bug:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/formula.rb:261:in `validate_attributes!'
/usr/local/Homebrew/Library/Homebrew/formula.rb:208:in `initialize'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `new'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `get_formula'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:331:in `factory'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:377:in `rescue in from_keg'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:373:in `from_keg'
/usr/local/Homebrew/Library/Homebrew/keg.rb:421:in `to_formula'
/usr/local/Homebrew/Library/Homebrew/keg.rb:135:in `block in find_some_installed_dependents'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `each'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `group_by'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `find_some_installed_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:112:in `check_for_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:105:in `handle_unsatisfied_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:44:in `uninstall'
/usr/local/Homebrew/Library/Homebrew/brew.rb:103:in `<main>'
  </pre>
</details>

## What you expected to happen

Formula removed.

**Comment by EricFromCanada** [comment]
This is a latent bug in _version.rb_ that was uncovered by #7091. Current behaviour:
```
$ brew tap riscv/riscv
==> Tapping riscv/riscv
Cloning into '/usr/local/Homebrew/Library/Taps/riscv/homebrew-riscv'...
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 467 (delta 4), reused 10 (delta 4), pack-reused 457
Receiving objects: 100% (467/467), 65.97 KiB | 2.27 MiB/s, done.
Resolving deltas: 100% (289/289), done.
Error: Invalid formula: /usr/local/Homebrew/Library/Taps/riscv/homebrew-riscv/riscv-gnu-toolchain.rb
invalid attribute for formula 'riscv/riscv/riscv-gnu-toolchain': version (nil)
Error: Invalid formula: /usr/local/Homebrew/Library/Taps/riscv/homebrew-riscv/riscv-isa-sim.rb
invalid attribute for formula 'riscv/riscv/riscv-isa-sim': version (nil)
Error: Invalid formula: /usr/local/Homebrew/Library/Taps/riscv/homebrew-riscv/riscv-pk.rb
invalid attribute for formula 'riscv/riscv/riscv-pk': version (nil)
Error: Cannot tap riscv/riscv: invalid syntax in tap!
```
This is happening because none of the listed formulae have a `version` attribute, nor a `url` containing a version number that's paresable by _version.rb_. After undoing the change to _version.rb_'s line 339:
```
$ brew tap riscv/riscv
==> Tapping riscv/riscv
Cloning into '/usr/local/Homebrew/Library/Taps/riscv/homebrew-riscv'...
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 467 (delta 4), reused 10 (delta 4), pack-reused 457
Receiving objects: 100% (467/467), 65.97 KiB | 2.36 MiB/s, done.
Resolving deltas: 100% (289/289), done.
Tapped 4 formulae (32 files, 109.2KB).

$ brew info riscv-gnu-toolchain | head -n1
riscv/riscv/riscv-gnu-toolchain: stable gnu
```
The regex `/-v?([^-]+)/` is (incorrectly) setting the version to "gnu" as it's the first word following the first dash in the URL. 

In short: before #7091, _version.rb_ wasn't flagging formula files with invalid syntax if the `url` contained a dash or an underscore.

**Comment by Bo98** [comment]
Indeed. I believe the version changes are correct and that the tap should be supplying `version` in this case, but ultimately it shouldn't leave the end user with an effectively broken Homebrew installation when situations like this happen.

**Comment by jabenninghoff** [comment]
I'm having a similar problem with my own tap: https://github.com/jabenninghoff/homebrew-edge

```
$ brew doctor
Error: invalid attribute for formula 'jabenninghoff/edge/radiosh': version (nil)
Please report this bug:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/formula.rb:261:in `validate_attributes!'
/usr/local/Homebrew/Library/Homebrew/formula.rb:208:in `initialize'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `new'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `get_formula'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:271:in `get_formula'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:331:in `factory'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:374:in `from_keg'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:350:in `from_rack'
/usr/local/Homebrew/Library/Homebrew/diagnostic.rb:714:in `block in check_for_unreadable_installed_formula'
/usr/local/Homebrew/Library/Homebrew/diagnostic.rb:713:in `each'
/usr/local/Homebrew/Library/Homebrew/diagnostic.rb:713:in `check_for_unreadable_installed_formula'
/usr/local/Homebrew/Library/Homebrew/cmd/doctor.rb:61:in `block in doctor'
/usr/local/Homebrew/Library/Homebrew/cmd/doctor.rb:53:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/doctor.rb:53:in `doctor'
/usr/local/Homebrew/Library/Homebrew/brew.rb:103:in `<main>'
```

To me this seems like a bug, although I was able to fix the issue by adding a `version` statement to the formula. It wasn't immediately obvious how to do this, perhaps updating the docs would help? The docs reference adding a statement, but not where...

**Comment by josharian** [comment]
Thanks for the speedy work! However, it doesn't look like #7109 fixed this for me.

```
$ brew update
Updated Homebrew from fab4a4863 to 2b838713a.
...
```

2b838713a contains #7109.

```
$ brew uninstall riscv-pk
Error: invalid attribute for formula 'riscv-pk': version (nil)
Please report this bug:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/formula.rb:261:in `validate_attributes!'
/usr/local/Homebrew/Library/Homebrew/formula.rb:208:in `initialize'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `new'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:122:in `get_formula'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:331:in `factory'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:377:in `rescue in from_keg'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:373:in `from_keg'
/usr/local/Homebrew/Library/Homebrew/keg.rb:421:in `to_formula'
/usr/local/Homebrew/Library/Homebrew/keg.rb:135:in `block in find_some_installed_dependents'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `each'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `group_by'
/usr/local/Homebrew/Library/Homebrew/keg.rb:132:in `find_some_installed_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:112:in `check_for_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:105:in `handle_unsatisfied_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/uninstall.rb:44:in `uninstall'
/usr/local/Homebrew/Library/Homebrew/brew.rb:103:in `<main>'
```

**Comment by MikeMcQuaid** [comment]
@josharian I'll wait for confirmation from others but regardless you'll need to fix your formula `version`.

**Comment by josharian** [comment]
> @josharian I'll wait for confirmation from others but regardless you'll need to fix your formula version.

I guess, but it's not my formula, and I haven't used it in years. All I really want to do is delete it. I would understand a lot more needing to fix it if I was trying to install it.

But I'll file an upstream issue for the formula later today.

**Comment by MikeMcQuaid** [comment]
Yeh, sorry, to be clear: I've reopened so I can fix the "can't uninstall" case.
